### PR TITLE
Feature/caas operator use svc ip

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

changes made for supporting `upgrade-[controller|model]` for k8s controller:
`apiaddresses` in agent config for a `CAAS` operator set to `CloudService.Addresses` for a k8s controller;

## QA steps

1. bootstrap into microk8s

```bash
$ juju bootstrap microk8s k1 --debug 
```

2. add model then deploy `CAAS` workloads;

```bash
$ juju add-model t1 microk8s && \
juju create-storage-pool mariadb-pv kubernetes storage-class=microk8s-hostpath && \
juju create-storage-pool operator-storage kubernetes storage-class=microk8s-hostpath && \
juju deploy cs:~juju/mariadb-k8s-0  --debug --storage database=500M,mariadb-pv && \
juju deploy cs:~juju/gitlab-k8s-0  --debug  --config juju-external-hostname=127.0.0.1.xip.io && \
juju relate mariadb-k8s gitlab-k8s && \
juju expose gitlab-k8s
```

3. verify `apiaddresses` in `agent.conf`;

```bash
$ kubectl exec gitlab-k8s-operator-0  -n t1 -it cat /var/lib/juju/agents/application-gitlab-k8s/agent.conf
```

## Documentation changes

None;

## Bug reference

Not ready;

### CODE REVIEW

File: environs/bootstrap/config.go [hunk 1]
```diff
@@ -241,7 +246,7 @@ var configChecker = schema.FieldMap(schema.Fields{
 	CACertKey + "-path":        schema.Omit,
 	CAPrivateKeyKey:            schema.Omit,
 	CAPrivateKeyKey + "-path":  schema.Omit,
-	ControllerServiceTypeKey:   schema.Omit,
+	ControllerServiceTypeKey:   DefaultControllerServiceType,
```
Comment (wallyworld) [inline environs/bootstrap/config.go]: I think it better to keep this as schema.Omit so we don't pollute IAAS config.
We can add the default later during the bootstrap config processing only if k8s controller.

File: state/address.go [hunk 2]
```diff
@@ -221,21 +221,71 @@ func (st *State) APIHostPortsForClients() ([][]network.HostPort, error) {
...
+// If the controller model is CAAS type, the return will be the controller
+// k8s service addresses in cloud service.
 // If there is no management network space configured for the controller,
 // or if the space is misconfigured, the return will be the same as
 // APIHostPortsForClients.
 // Otherwise the returned addresses will correspond with the management net space.
 // If there is no document at all, we simply fall back to APIHostPortsForClients.
 func (st *State) APIHostPortsForAgents() ([][]network.HostPort, error) {
-	hp, err := st.apiHostPortsForKey(apiHostPortsForAgentsKey)
+	if hps, err := st.apiHostPortsForOperator(); err == nil {
```
Comment (wallyworld) [inline state/address.go]: We are ignoring potential fatal errors.
We need instead to get isCAAS from the model and call the correct method

File: state/address.go [hunk 3]
```diff
@@ -221,21 +221,71 @@ func (st *State) APIHostPortsForClients() ([][]network.HostPort, error) {
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	apiPort := controllerConfig.APIPort()
+	svc, err := controllerSt.CloudService(controllerUUID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return [][]network.HostPort{
+		network.AddressesWithPort(svc.Addresses(), apiPort),
```
Comment (wallyworld) [inline state/address.go]: We need to filter addresses to only get those which are of type CloudLocal
We discussed that we could store in []addresses Public and CloudLocal and so we need to filter. If there's only one, then we use that regardless.

File: environs/bootstrap/config.go (hunk 1 above)
Comment (ycliuhw) [reply environs/bootstrap/config.go]: ah, right

File: state/address.go (hunk 2 above)
Comment (ycliuhw) [reply state/address.go]: ic

File: state/address.go (hunk 3 above)
Comment (ycliuhw) [reply state/address.go]: I thought it would be next PR. 
no problem, it's address related. I am adding this functionality now

**Comment by ycliuhw** [comment]
[Juju CLI should talk with k8s controller via Service DNS rather than pod IP](https://trello.com/c/K03gIYQj/674-juju-cli-should-talk-with-k8s-controller-via-service-dns-rather-than-pod-ip)

**Comment by ycliuhw** [comment]
!!build!!

File: cloudconfig/podcfg/podcfg.go [hunk 4]
```diff
@@ -265,7 +265,11 @@ func (cfg *ControllerPodConfig) verifyBootstrapConfig() (err error) {
 		return errors.New("entity tag must be nil when bootstrapping")
 	}
 	if cfg.Bootstrap.ControllerServiceType == "" {
-		return errors.New("controller service type is required")
+		return errors.New(`
+controller service type is missing.
+The k8s cloud might be staled, please try to re-import k8s cloud using
```
Comment (wallyworld) [inline cloudconfig/podcfg/podcfg.go]:
```
The k8s cloud definition might be stale, please to try reiimport the k8s cloud using
     juju add-k8s <cloud-name> --cluster-name <cluster-name> --local

See juju help add-k8s for more information.
```

File: state/address.go [hunk 5]
```diff
@@ -221,21 +230,82 @@ func (st *State) APIHostPortsForClients() ([][]network.HostPort, error) {
...
 			return st.APIHostPortsForClients()
 		}
 		return nil, errors.Trace(err)
 	}
-	return hp, nil
+	return hps, nil
+}
+
+func (st *State) isCAASController() (bool, error) {
+	ctrlSt, err := st.newStateNoWorkers(st.ControllerModelUUID())
```
Comment (wallyworld) [inline state/address.go]: This is much easier done like this:

```
	ctrlModel := &Model{
		st: st,
	}
	if err := ctrlModel.refresh(st.ControllerModelUUID()); err != nil {
		return false, errors.Trace(err)
	}
        return ctrlModel.Type() == ModelTypeCAAS, nil

```

File: cloudconfig/podcfg/podcfg.go (hunk 4 above)
Comment (ycliuhw) [reply cloudconfig/podcfg/podcfg.go]: yup, changed

File: state/address.go (hunk 5 above)
Comment (ycliuhw) [reply state/address.go]: changed

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**